x/metadata/types/address.go

```diff
@@ -318,35 +355,77 @@ func (ma MetadataAddress) Compare(other MetadataAddress) int {
 
 // ScopeUUID returns the scope uuid component of a MetadataAddress (if appropriate)
 func (ma MetadataAddress) ScopeUUID() (uuid.UUID, error) {
+	if !ma.isTypeOneOf(ScopeKeyPrefix, GroupKeyPrefix, RecordKeyPrefix) {
+		return uuid.UUID{}, fmt.Errorf("this metadata address does not contain a scope uuid")
+	}
+	return ma.PrimaryUUID()
+}
+
+// GroupUUID returns the group uuid component of a MetadataAddress (if appropriate)
+func (ma MetadataAddress) GroupUUID() (uuid.UUID, error) {
+	if len(ma) > 0 && ma[0] != GroupKeyPrefix[0] {
+		return uuid.UUID{}, fmt.Errorf("this metadata address does not contain a group uuid")
+	}
+	return ma.SecondaryUUID()
+}
+
+func (ma MetadataAddress) ContractSpecUUID() (uuid.UUID, error) {
+	if !ma.isTypeOneOf(ContractSpecificationKeyPrefix, RecordSpecificationKeyPrefix) {
+		return uuid.UUID{}, fmt.Errorf("this metadata address does not contain a contract specification uuid")
+	}
+	return ma.PrimaryUUID()
+}
+
+// PrimaryUUID returns the primary UUID from this MetadataAddress (if applicable).
+// More accurately, this converts bytes 2 to 17 to a UUID.
+// For example, if this MetadataAddress is for a scope specification, this will return the scope specification uuid.
+// But if this MetadataAddress is for a record specification, this will return the contract specification
+// (since that's the first part of those metadata addresses).
+func (ma MetadataAddress) PrimaryUUID() (uuid.UUID, error) {
 	if len(ma) < 1 {
 		return uuid.UUID{}, fmt.Errorf("address empty")
 	}
 	// if we don't know this type
-	if ma[0] > 0x04 {
+	if !ma.isTypeOneOf(allowedTypes...) {
 		return uuid.UUID{}, fmt.Errorf("invalid address type out of valid range (got: %d)", ma[0])
 	}
-	// Scope, RecordGroup, Record all have a ScopeUUID
-	if ma[0] == ScopeKeyPrefix[0] || ma[0] == GroupKeyPrefix[0] || ma[0] == RecordKeyPrefix[0] {
-		return uuid.FromBytes(ma[1:17])
+	if len(ma) < 17 {
+		return uuid.UUID{}, fmt.Errorf("incorrect address length (must be at least 17, actual: %d)", len(ma))
 	}
-	// otherwise there isn't a scope uuid so return empty.
-	return uuid.UUID{}, fmt.Errorf("this metadata address does not contain a scope uuid")
+	return uuid.FromBytes(ma[1:17])
 }
 
-// GroupUUID returns the group uuid component of a MetadataAddress (if appropriate)
-func (ma MetadataAddress) GroupUUID() (uuid.UUID, error) {
+// SecondaryUUID returns the secondary UUID from this MetadataAddress (if applicable).
+// More accurately, this converts bytes 18 to 33 (inclusive) to a UUID.
+func (ma MetadataAddress) SecondaryUUID() (uuid.UUID, error) {
 	if len(ma) < 1 {
 		return uuid.UUID{}, fmt.Errorf("address empty")
 	}
 	// if we don't know this type
-	if ma[0] > 0x04 {
+	if !ma.isTypeOneOf(allowedTypes...) {
```

Comment:
Not all of these have a second uuid (even when they have a length that is greater than 33).  When written the only MetadataAddress type with two uuids was the RecordGroup (Session) which indicates an object that is part of a scope and has its own uuid.  If the intention is to have a generic second uuid method which might be used by other id types (which?) then it should be scoped private since it currently doesn't provide appropriate context checks.

Reply:
Good catch. The change I made reflected the previous functionality, but that wasn't exactly accurate either. Should be fixed now.